File: server/stackexchange/hs/src/Webar/Server/StackExchange/Source.hs
{-# LANGUAGE TemplateHaskell #-}

module Webar.Server.StackExchange.Source where

import Webar.Data.TH
import Webar.Object
import qualified Webar.Server.StackExchange.Image.Source as Img.Source
import qualified Webar.Server.StackExchange.RestApi.Source as RestApi
import Webar.Types (Server (..), Version (Version))

server :: Server
server = Server {serverName = "StackExchange", serverVersion = Version 1 0}

data FetchType
  = FtRestApi
  | FtImage
  deriving (Show, Eq)

deriveSumData
  defaultSumOptions {constructorTagModifier = camelTo2 '_' . drop 2}
  ''FetchType

data RecordType
  = RtRestApi {-# UNPACK #-} RestApi.RequestRecord
  | RtFilter
  | RtImageRequest {-# UNPACK #-} Img.Source.RequestRecord
  deriving (Show, Eq)

deriveSumData
  defaultSumOptions
    { constructorTagModifier = camelTo2 '_' . drop 2
    }
  ''RecordType

data ArchiveInfo
  = AiRestApi RestApi.Archive
  | AiImage Img.Source.ArchiveImage
  deriving (Show, Eq)

deriveSumData
  defaultSumOptions
    { constructorTagModifier = camelTo2 '_' . drop 2
    }
  ''ArchiveInfo

type ArchiveId = ObjectId ArchiveInfo

data SnapshotType
  = StRestApi RestApi.SnapshotType
  | StImage {-# UNPACK #-} Img.Source.SnapshotType
  deriving (Show, Eq)

deriveSumData
  defaultSumOptions
    { constructorTagModifier = camelTo2 '_' . drop 2
    }
  ''SnapshotType

File: server/stackexchange/core-rs/src/source.rs
use serde::Deserialize;

use webar_core::{object::Server, Version};
use webar_data::ser::Serialize;

use crate::image;

pub const SERVER: Server<&'static str> = Server {
    name: "StackExchange",
    version: Version(1, 0),
};

pub type Instance = ();

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum FetchType {
    #[serde(rename = "rest_api")]
    RestApi,
    #[serde(rename = "image")]
    Image,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum RecordType {
    #[serde(rename = "image_request")]
    ImageRequest(image::source::RequestRecord),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum ArchiveInfo<S> {
    #[serde(rename = "image")]
    Image(image::source::ArchiveImage<S>),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum SnapshotType {
    #[serde(rename = "image")]
    Image(image::source::SnapshotType),
}
